geode-gfsh/src/main/java/org/apache/geode/management/internal/cli/functions/WanCopyRegionFunction.java

```diff
@@ -0,0 +1,480 @@
+          return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+              CliStrings.format(
+                  CliStrings.WAN_COPY_REGION__MSG__CANCELED__AFTER__HAVING__COPIED,
+                  copiedEntries));
+        }
+      }
+    } finally {
+      if (senderPool != null && connection != null) {
+        ((PooledConnection) connection).setShouldDestroy();
+        senderPool.returnConnection(connection);
+      }
+    }
+
+    if (region.isDestroyed()) {
+      return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.ERROR,
+          CliStrings.format(
+              CliStrings.WAN_COPY_REGION__MSG__ERROR__AFTER__HAVING__COPIED,
+              "Region destroyed",
+              copiedEntries));
+    }
+
+    return new CliFunctionResult(context.getMemberName(), CliFunctionResult.StatusState.OK,
+        CliStrings.format(CliStrings.WAN_COPY_REGION__MSG__COPIED__ENTRIES,
+            copiedEntries));
+  }
+
+  List<GatewayQueueEvent> createBatch(InternalRegion region, GatewaySender sender,
+      int batchSize, InternalCache cache, Iterator<?> iter) {
+    int batchIndex = 0;
+    List<GatewayQueueEvent> batch = new ArrayList<>();
+    while (iter.hasNext() && batchIndex < batchSize) {
+      GatewayQueueEvent event =
+          createGatewaySenderEvent(cache, region, sender, (Region.Entry) iter.next());
+      if (event != null) {
+        batch.add(event);
+        batchIndex++;
+      }
+    }
+    return batch;
+  }
+
+  Set<?> getEntries(Region region, GatewaySender sender) {
+    if (region instanceof PartitionedRegion && sender.isParallel()) {
+      return ((PartitionedRegion) region).getDataStore().getAllLocalBucketRegions()
+          .stream()
+          .flatMap(br -> ((Set<?>) br.entrySet()).stream()).collect(Collectors.toSet());
+    }
+    return region.entrySet();
+  }
+
+  @VisibleForTesting
+  GatewayQueueEvent createGatewaySenderEvent(InternalCache cache,
+      InternalRegion region, GatewaySender sender, Region.Entry entry) {
+    final EntryEventImpl event;
+    if (region instanceof PartitionedRegion) {
+      event = createEventForPartitionedRegion(sender, cache, region, entry);
+    } else {
+      event = createEventForReplicatedRegion(cache, region, entry);
+    }
+    if (event == null) {
+      return null;
+    }
+    try {
+      return new GatewaySenderEventImpl(EnumListenerEvent.AFTER_UPDATE_WITH_GENERATE_CALLBACKS,
+          event, null, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  final CliFunctionResult cancelWanCopyRegion(FunctionContext<Object[]> context,
+      String regionName, String senderId) {
+    boolean found = false;
+    String threadBaseName = getFunctionThreadName(regionName, senderId);
+    for (Thread t : Thread.getAllStackTraces().keySet()) {
+      if (t.getName().startsWith(threadBaseName)) {
+        found = true;
+        t.interrupt();
```

Comment:
The problem is, where would that list of threads be kept? It could be in a static variable but it does not look like the cleanest solution.
If we open the door to using static variables to identify the command to cancel we could also explore a solution I considered in the past that did not require to run the copy region code in a new thread in order to be able to cancel it later. We could have a static variable that would be used to specify an execution to be canceled (identified by senderId and region) and the function could check after each batch is sent if that variable indicates that the function should be canceled.
This way the identification of the command to cancel would be more efficient than what we are doing currently and we would get rid of the new thread to execute the code.

Anyway, I am not sure which alternative is better.

Reply:
It will not be the cleanest solution, the idea that you can start stuff from the command line that take time is kind of challenge. If a shutdown of the system happens or a close cache, you are going to want a graceful way to shut this down. From my perspective it would be a protected list of "Threads in progress". I  would be interested to hear what others think as I have not done something like this in this space either.

Reply:
@upthewaterspout  Do you have any thoughts?

Reply:
@mhansonp  Who would be stopping these executions? If it must be done from outside the function when the system is going to be shutdown or the cache to be closed, we could just find all the threads whose name starts with a given prefix (wanCopyRegionFunctionThread_) and send the interrupt signal to them.
That way you would not need to keep a list of threads in progress for this command.

Reply:
That is sufficient. I much rather bear the expense of a couple of Thread objects than a lookup like this, but I am fine with it being a lookup. Other than that, things look good.

Reply:
If you can add someway to cleanly shut this down, I think you have all the necessary bits stuff and the code looks good.
I think it is important to consider what should happen with a partial transfer on the local and the remote when a shutdown or close happens...  I would suggest throwing away the data and finishing the command as a first step others might disagree.

Reply:
I have added a change that allows to stop all executions by passing "*" as region name and "*" as senderId to the function.

Reply:
I think an automatic approach is really needed. you cannot connect to geode if it is shutting down go run the command.

Reply:
this should upon checking cancellation and getting a cancel automatically terminate the copy. This assumes that we care if there is a partial copy which I assume we do.

Reply:
@mhansonp I am not sure I understand what you mean.
Would it be ok to check inside the function if there is a cancel in progress, and if so, terminate it, reporting in the output that it was terminated due to the cache being closed?
For example, add the following at the beginning of `doPostSendBatchActions()`:

```
if (cache.getCancelCriterion().isCancelInProgress()) {
<terminate function and report in the output of the command that the cache was closed>
}
```

Reply:
Let me check with some of my peers.